File: app/src/main/java/com/workingagenda/democracydroid/MainActivity.java
/*
 *
 *   Copyright (C) 2014-2015 Fenimore Love
 *
 *   This program is free software: you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation, either version 3 of the License, or
     (at your option) any later version.
 *   This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.workingagenda.democracydroid;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.preference.PreferenceManager;
import androidx.viewpager2.adapter.FragmentStateAdapter;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.workingagenda.democracydroid.databinding.ActivityMainBinding;
import com.workingagenda.democracydroid.tabfragment.DownloadFragment;
import com.workingagenda.democracydroid.tabfragment.PodcastFragment;
import com.workingagenda.democracydroid.tabfragment.StoryFragment;

public class MainActivity extends AppCompatActivity {
    private static final int POS_STORY = 0;
    private static final int POS_PODCAST = 1;
    private static final int POS_DOWNLOAD = 2;
    private static final int TOTAL_COUNT = 3;
    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.mainToolbar);
        // Shared Preferences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        int DEFAULT_TAB = Integer.parseInt(preferences.getString("pref_default_tab", "0"));
        boolean PREF_FIRST_TIME = preferences.getBoolean("first_preference", true);
        // TODO: have splash screen for new users
        Log.d("First time", String.valueOf(PREF_FIRST_TIME));

        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        final SectionsStateAdapter mSectionsStateAdapter = new SectionsStateAdapter(this);

        // Set up the ViewPager with the sections adapter.
        binding.mainViewPager2.setAdapter(mSectionsStateAdapter);

        new TabLayoutMediator(binding.mainTabLayout, binding.mainViewPager2,
                mSectionsStateAdapter::getPageIcon
        ).attach();
        binding.mainTabLayout.setTabGravity(TabLayout.GRAVITY_FILL);
        binding.mainViewPager2.setCurrentItem(DEFAULT_TAB);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        switch (id) {
            case R.id.menu_main_settings:
                Intent intent1 = new Intent(this, SettingsActivity.class);
                startActivity(intent1);
                return true;
            case R.id.menu_main_refresh:
                // Don't let user click before async tasks are done
                item.setEnabled(false);
                // Call Fragment refresh methods
                getSupportFragmentManager().getFragments();
                for (Fragment x : getSupportFragmentManager().getFragments()) {
                    if (x instanceof PodcastFragment)
                        ((PodcastFragment) x).refresh();
                    if (x instanceof StoryFragment)
                        ((StoryFragment) x).refresh();
                }
                // FIXME: Somehow enable this after async call...
                item.setEnabled(true);
                return true;
            case R.id.menu_main_exclusives:
                String url = "https://www.democracynow.org/categories/web_exclusive";
                Intent i = new Intent(Intent.ACTION_VIEW);
                i.setData(Uri.parse(url));
                startActivity(i);
                return true;
            case R.id.menu_main_site:
                String url1 = "http://www.democracynow.org/";
                Intent i1 = new Intent(Intent.ACTION_VIEW);
                i1.setData(Uri.parse(url1));
                startActivity(i1);
                return true;
            case R.id.menu_main_about:
                Intent intent = new Intent(this, AboutActivity.class);
                startActivityForResult(intent, 0);
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * A {@link FragmentStateAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    private static class SectionsStateAdapter extends FragmentStateAdapter {

        SectionsStateAdapter(final FragmentActivity fa) {
            super(fa);
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            // getItem is called to instantiate the fragment for the given page.
            switch (position) {
                default: // Deliberate fall-through to story fragment
                case POS_STORY:
                    return new StoryFragment();
                case POS_PODCAST:
                    return new PodcastFragment();
                case POS_DOWNLOAD:
                    return new DownloadFragment();
            }
        }

        void getPageIcon(final TabLayout.Tab tab, final int position) {
            switch (position) {
                default: // Deliberate fall-through to story fragment
                case POS_STORY:
                    tab.setIcon(R.drawable.ic_library_books);
                    tab.setContentDescription(R.string.tab_transcripts);
                    break;
                case POS_PODCAST:
                    tab.setIcon(R.drawable.ic_live_tv);
                    tab.setContentDescription(R.string.tab_broadcasts);
                    break;
                case POS_DOWNLOAD:
                    tab.setIcon(R.drawable.ic_download_white);
                    tab.setContentDescription("Downloads");
            }
        }

        @Override
        public int getItemCount() {
            // Show 2 total pages.
            return TOTAL_COUNT;
        }
    }
}

File: app/src/main/java/com/workingagenda/democracydroid/ui/main/MainActivity.kt
/*
 *  Copyright (C) 2014-2015 Democracy Droid
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.workingagenda.democracydroid.ui.main

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import butterknife.BindView
import butterknife.ButterKnife
import com.workingagenda.democracydroid.R
import com.workingagenda.democracydroid.ui.FragmentRefreshListener
import com.workingagenda.democracydroid.ui.about.AboutActivity
import com.workingagenda.democracydroid.ui.download.DownloadFragment
import com.workingagenda.democracydroid.ui.feed.FeedFragment
import com.workingagenda.democracydroid.ui.feed.FeedType
import com.workingagenda.democracydroid.ui.settings.SettingsActivity
import com.workingagenda.democracydroid.util.SharedPreferenceManager

class MainActivity : AppCompatActivity() {

    @BindView(R.id.appbar_layout) lateinit var appbarLayout: AppBarLayout
    @BindView(R.id.toolbar) lateinit var toolbar: Toolbar
    @BindView(R.id.tab_layout) lateinit var tabLayout: TabLayout
    @BindView(R.id.container) lateinit var viewPager: ViewPager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        setupTabLayout()
        setupViewPager()
        setupAdapter()
        setSupportActionBar(toolbar)
    }

    private fun setupTabLayout(){
        tabLayout.addTab(tabLayout.newTab().setIcon(R.drawable.ic_library_books_white_24dp))
        tabLayout.addTab(tabLayout.newTab().setIcon(R.drawable.ic_live_tv_white_24dp))
        tabLayout.addTab(tabLayout.newTab().setIcon(R.drawable.ic_file_download_white_24dp))
        tabLayout.tabGravity = TabLayout.GRAVITY_FILL
    }

    private fun setupViewPager(){
        viewPager.offscreenPageLimit = 2
        viewPager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tabLayout))
    }

    private fun setupAdapter(){
        viewPager.adapter = SectionsPagerAdapter(supportFragmentManager)
        tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                viewPager.currentItem = tab.position
                appbarLayout.setExpanded(true,true)
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {
            }

            override fun onTabReselected(tab: TabLayout.Tab) {
            }
        })
        viewPager.currentItem = SharedPreferenceManager.getTabPreference()
    }

    private fun actionViewIntent(url:String) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(url)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean{
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean{
        val id = item.itemId
        when(id){
            R.id.action_settings -> {
                val intent = Intent(this, SettingsActivity::class.java)
                startActivity(intent)
                return true
            }
            R.id.action_refresh -> {
                item.isEnabled = false
                for (x in supportFragmentManager.fragments) {
                    (x as? FragmentRefreshListener)?.refresh()
                }
                item.isEnabled = true
                return true
            }
            R.id.action_donate -> actionViewIntent("https://www.democracynow.org/donate")
            R.id.action_exclusives -> actionViewIntent("https://www.democracynow.org/categories/web_exclusive")
            R.id.action_site -> actionViewIntent("http://www.democracynow.org/")
            R.id.action_about -> {
                val intent = Intent(this, AboutActivity::class.java)
                startActivityForResult(intent, 0)
            }
        }
        return true
    }

    inner class SectionsPagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {

        override fun getItem(position: Int): Fragment = when (position) {
            0 -> FeedFragment.newInstance(FeedType.STORY)
            1 -> FeedFragment.newInstance(FeedType.EPISODE)
            2 -> DownloadFragment()
            else -> FeedFragment.newInstance(FeedType.EPISODE)
        }

        override fun getCount(): Int = 3
    }
}

